marshmallow/marshalling.py

```diff
@@ -250,57 +250,55 @@ def deserialize(
             return ret
         partial_is_collection = is_collection(partial)
         ret = dict_class()
-        for attr_name, field_obj in iteritems(fields_dict):
-            if field_obj.dump_only:
-                continue
-            field_name = attr_name
-            if field_obj.data_key:
-                field_name = field_obj.data_key
-            try:
+        # Check data is a dict
+        try:
+            data.get('dummy', missing)
```

Comment:
I think doing `if isinstance(data, collections.Mapping)` is more appropriate in this case.